### PR TITLE
Support for `dtype` keyword in tf.layers.*

### PR DESCRIPTION
**Pull request by lakshayg**
This pull request introduces the `dtype` keyword in functional interfaces for core layers to solve #9898. All the unit-tests seem to run fine on my machine after incorporating the changes. The default value has been kept as tf.float32 in order to ensure compatibility.

### CODE REVIEW

**Comment by fchollet** [comment]
We have merged yesterday a change that allows fp16/64 support in all core layers where it applies.

**Comment by jhseu** [comment]
Ah, yeah, closing for now. The open-source code should have those dtype updates when I push later today.

**Comment by yangjunpro** [comment]
As far as I know, for different precision-level dtypes, there are some tricks taking into consideration.
For example, for fp16, the loss may needs to be adjusted to compensate for the potential accuracy degradation.  
So I think a straight-forward template-based data type enhancement may not be enough.

**Comment by harpone** [comment]
So what's the situation with this now? Not seeing any `dtype` kwargs (in conv layers at least)...

Is it safe to copy and modify the conv layers to (optionally) use `float16`?

**Comment by rockclimber112358** [comment]
@yangjunpro Providing this functionality would still be useful.  As always, it's up to the user to verify that a change to the model doesn't hurt performance, and many models don't require loss scaling to still be effective with float16 (https://devblogs.nvidia.com/mixed-precision-training-deep-neural-networks/).
